emma-language/src/main/scala/eu/stratosphere/emma/compiler/lang/backend/Order.scala

```diff
@@ -0,0 +1,172 @@
+ * we will later transform higher-order code differently than first-order code.
+ *
+ * Note: This code assumes that DataBags don't contain functions. (E.g. it handles a DataBag[Int => Int] incorrectly.)
+ */
+private[backend] trait Order extends Common {
+  self: Backend with Core =>
+
+  import Core.{Lang => core}
+  import UniverseImplicits._
+
+  private[backend] object Order {
+
+    /**
+     * If a lambda is given as an argument to one of these methods,
+     * then that lambda will be called from higher-order context.
+     */
+    val combinators = API.methods ++ ComprehensionCombinators.methods
+
+    /**
+     * Disambiguates order in a tree and gives information on which parts of the code might be executed
+     * from a higher-order context.
+     *
+     * == Preconditions ==
+     *
+     * - Input must be in ANF.
+     * - DataBags don't contain functions. (TODO: Maybe add a check for this in some place like CoreValidate)
+     *
+     * == Postconditions ==
+     *
+     * Returns a (disambiguatedTree, isHighContext) pair, where
+     * disambiguatedTree is an ANF tree where it has been decided for every piece of code whether to treat it
+     * as driver-only or high (that is, it can be called from a higher-order context). (Some ValDefs have been
+     * duplicated for this.)
+     *
+     * isHighContext can be used as an inherited attribute to know which code might be called from higher-order context
+     * (e.g. like this: api.BottomUp.inherit(isHighContext)(Monoids.disj).transform{...}(disambiguatedTree)).
+     * Note, that isHighContext might contain false positives, i.e. it might give true for code that is only executed
+     * in the driver.
+     *
+     * Warning: isHighContext depends on information about all the lambdas in disambiguatedTree and looking at the
+     * combinator calls. Therefore, it will not work on a tree which was got from disambiguatedTree by changing
+     * the lambdas or the combinator calls. (Note: this is why the backend transformations have to go BottomUp.)
+     *
+     * Algorithm:
+     * 1. Collect symbols of all function ValDefs. (funs)
+     * 2. Build a graph of funs referencing funs. (funGraph)
+     * 3. Collect funs that are given as arguments to combinators. (topLevRefs)
+     * 4. Collect funs referenced from top-level. (combRefs)
+     * 5. Do graph traversals on funGraph from combRefs and topLevRefs to get all high and
+     *    driver-only funs, respectively. (highFuns0, driverFuns)
+     * 6. The intersection of highFuns0 and driverFuns are the ambiguous ones. Create new ValDefs
+     *    for these, and modify refs to them in high code to the newly created vals.
+     */
+    lazy val disambiguate: u.Tree => (u.Tree, u.Tree =?> Boolean) = tree => {
+
+      def isFun(sym: u.TermSymbol) = api.Sym.funs(api.Type.of(sym).typeSymbol)
+
+      // The Set of symbols of ValDefs of functions
+      val funs = tree.collect {
+        case api.ValDef(sym, _, _) if isFun(sym) => sym
```

Comment:
should be `core.ValDef` (we try to not use the `api.*` objects explicitly if the logic can be defined on top of the `core.*` language.